File: c/validators/writer/test_uv_back.c
// SPDX-License-Identifier: MIT
#include "test_utils.h"

static bool validate_uv_back(const char* directory) {
    const char* fixture = "uv_back";
    SUModelRef model = SU_INVALID;
    bool opened = false;
    bool valid = false;
    SUEntitiesRef entities = SU_INVALID;
    SUFaceRef face = SU_INVALID;
    SUMaterialRef material = SU_INVALID;
    SUMaterialRef back_material = SU_INVALID;
    char name[NAME_CAPACITY] = {0};

    opened = open_fixture(fixture, directory, "uv_back.skp", &model);
    if (!opened) return false;
    REQUIRE_SDK(fixture, SUModelGetEntities(model, &entities));
    REQUIRE(fixture, first_face(entities, &face), "could not retrieve two-sided face");
    REQUIRE_SDK(fixture, SUFaceGetFrontMaterial(face, &material));
    REQUIRE(fixture, material_name(material, name),
            "could not read front material name");
    REQUIRE(fixture, strcmp(name, "Writer Front") == 0,
            "front face material was not preserved");
    REQUIRE_SDK(fixture, SUFaceGetBackMaterial(face, &back_material));
    REQUIRE(fixture, material_name(back_material, name),
            "could not read back material name");
    REQUIRE(fixture, strcmp(name, "Writer Back") == 0,
            "back face material was not preserved");
    valid = true;

cleanup:
    if (opened) SUModelRelease(&model);
    return valid;
}


WRITER_VALIDATOR_MAIN("uv_back", validate_uv_back)

File: python/writer/generators/test_uv_back.py
# SPDX-License-Identifier: MIT
from pathlib import Path

from .common import run


def generate(skppy, destination: Path) -> None:
    model = skppy.new_model()
    front = model.add_material("Writer Front", color=skppy.Color(190, 20, 30))
    back = model.add_material("Writer Back", color=skppy.Color(30, 40, 190))
    face = model.entities.add_face([(0, 0, 0), (2, 0, 0), (0, 2, 0)])
    face.front_material_id = front.id
    face.back_material_id = back.id
    skppy.save(model, destination)


if __name__ == "__main__":
    raise SystemExit(run(generate, "uv_back"))
